### PR TITLE
middleware/dnstap add

### PR DESCRIPTION
**Pull request by 2buisson**
Add a dnstap middleware, not to be merged!

### CODE REVIEW

**Comment by johnbelamaric** [comment]
By convention we stick `WIP` (for work-in-progress) in front of PRs that we don't want someone to merge yet.

**Comment by miekg** [comment]
I will review this on <strike>Saturday (or maybe tomorrow).</strike>now

File: middleware/dnstap/handler.go
Comment (miekg) [inline middleware/dnstap/handler.go:3]: these import need to be sorted:
* stdlib, newline
* coredns ones, newline
* third party ones

File: middleware/dnstap/handler.go
Comment (miekg) [inline middleware/dnstap/handler.go]: rename lib to dnstap or tap

File: middleware/dnstap/handler.go
Comment (miekg) [inline middleware/dnstap/handler.go]: catch -> taprw ?

File: middleware/dnstap/handler.go
Comment (miekg) [inline middleware/dnstap/handler.go]: Why do you take the address of catch here?

File: middleware/dnstap/msg/msg.go
Comment (miekg) [inline middleware/dnstap/msg/msg.go]: Needs a comment

File: middleware/dnstap/msg/msg.go
Comment (miekg) [inline middleware/dnstap/msg/msg.go]: Would rename this to `epoch`

File: middleware/dnstap/msg/msg.go
Comment (miekg) [inline middleware/dnstap/msg/msg.go]: Why errors.Wrap?

File: middleware/dnstap/msg/msg.go
Comment (miekg) [inline middleware/dnstap/msg/msg.go]: almost identical to now, except for the `*dest = &buf`, should be possible to life without this.

File: middleware/dnstap/msg/msg.go
Comment (miekg) [inline middleware/dnstap/msg/msg.go]: socket -> family or socketFamily?

File: middleware/dnstap/msg/msg.go
Comment (miekg) [inline middleware/dnstap/msg/msg.go]: I think its nicer to *not* give `m *Msg` to this function, but let this return the things you're interested in and then assign the things you need to the fieds of m?

File: middleware/dnstap/msg/msg.go
Comment (miekg) [inline middleware/dnstap/msg/msg.go]: why do you need ipPort?

Comment (miekg) [reply middleware/dnstap/msg/msg.go]: You should be able to use a lot of stuff from `request.Request` and if not we should extend that package a little.

File: middleware/dnstap/msg/wrapper.go
Comment (miekg) [inline middleware/dnstap/msg/wrapper.go]: Better named `New`?

File: middleware/dnstap/setup.go
Comment (miekg) [inline middleware/dnstap/setup.go]: I wonder how useful this is as opposed to a socket. Also this name should be configurable as having a default name that a process with root access creates can lead to security issues.

File: middleware/dnstap/setup.go
Comment (miekg) [inline middleware/dnstap/setup.go]: This closes to above started go-routine?

File: middleware/dnstap/writer.go
Comment (miekg) [inline middleware/dnstap/writer.go]: We usually send state between function as-is, not as a pointer.

**Review by miekg — COMMENTED** [review]
Starting to look good. I've left a whole bunch of comments. Don't feel intimidated, if something wasn't clear I left a comment or question, even if it was a small nit.

**Comment by 2buisson** [comment]
I did a lot of changes since then, I will try to follow your directives before pushing them, thanks.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: middlewa..." ]
>I did a lot of changes since then, I will try to follow your directives before pushing them, thanks.

Ack. Overall looks good. Let us know when it is ready for a review.

/Miek

--
Miek Gieben

**Comment by 2buisson** [comment]
Here it is:
* I am wondering if this is the best approch: [writer.go](https://github.com/varyoo/coredns/blob/919bd9dd67d987b4f60cdcf6da2f4ad9442e8a92/middleware/dnstap/writer.go#L28), should it rather be done before the response is written back to the client, or asynchronously (a lot more of issues here).
* dnstap database VS socket: [README.md](https://github.com/varyoo/coredns/blob/919bd9dd67d987b4f60cdcf6da2f4ad9442e8a92/middleware/dnstap/README.md)

**Comment by miekg** [comment]
you should do it synchronise. The goroutine itself runs async so there is not need to do this again.

I also prefer the socket option. But you're saying you can't read while we write? I don't fully understand that.

**Comment by johnbelamaric** [comment]
If it's synchronous it will add some latency, I would think. But perhaps not enough to matter.

**Comment by 2buisson** [comment]
When the connection allows reading and writing, a bidirectional framestream can be used: before the client sends its data, the protocol ensure that the other party can read this data, by some read/write on both sides.

When using a socket the dnstap tool expects a bidirectional framestream. Only the bidirectional framestream decoder has been ported from the C library to the Go one, the encoder isn't there yet https://github.com/farsightsec/golang-framestream/pull/4.

**Comment by miekg** [comment]
Ack. Thanks for the update.

File: middleware/dnstap/README.md
Comment (alkino) [inline middleware/dnstap/README.md]: huge

File: middleware/dnstap/README.md
Comment (alkino) [inline middleware/dnstap/README.md]: "cannot be cleared easily while the server is running"?

File: middleware/dnstap/README.md
Comment (alkino) [inline middleware/dnstap/README.md]: doesn't

**Comment by johnbelamaric** [comment]
@varyoo looks like you have made some progress on https://github.com/farsightsec/golang-framestream/pull/4 - nice! What's the next step?

**Comment by 2buisson** [comment]
I would appreciate a review at this point before I start writing tests.

File: middleware/dnstap/README.md
Comment (miekg) [inline middleware/dnstap/README.md]: This README.md should follow the spec laid out in middleware.md

File: middleware/dnstap/README.md
Comment (miekg) [inline middleware/dnstap/README.md]: Is this still current? What does `true` mean?

File: middleware/dnstap/handler.go
Comment (miekg) [inline middleware/dnstap/handler.go]: Go doesn't have a line width, just put this on 1 line

File: middleware/dnstap/handler.go
Comment (miekg) [inline middleware/dnstap/handler.go]: `QueryEpoch()` seems to *set* the epoch value?

File: middleware/dnstap/handler.go
Comment (miekg) [inline middleware/dnstap/handler.go]: Just a nit on the name, but there is also a ServerTap? Otherwise just name this Tap

File: middleware/dnstap/writer.go
Comment (miekg) [inline middleware/dnstap/writer.go]: Tap -> Taper

File: middleware/dnstap/writer.go
Comment (miekg) [inline middleware/dnstap/writer.go]: This method should not log by itself , but just return the error to the caller.

File: middleware/dnstap/writer.go
Comment (miekg) [inline middleware/dnstap/writer.go]: I don't see an immediate need for errors.Wrap here? (or in general)

File: middleware/dnstap/writer.go
Comment (miekg) [inline middleware/dnstap/writer.go]: What does `Pack` mean?

Comment (miekg) [reply middleware/dnstap/writer.go]: Ah, I see, you need that binary for dnstap.

File: middleware/dnstap/msg/msg.go
Comment (miekg) [inline middleware/dnstap/msg/msg.go]: Are you used to a lot of C? In Go you should just return the multiple values instead of giving the function a pointer which it modifies. Why can't this function just return the `net.Addr` and let the caller deal with that?

File: middleware/dnstap/msg/msg.go
Comment (miekg) [inline middleware/dnstap/msg/msg.go]: might want to make this explicit instead of `default` and leave `default` when nothing matches.

File: middleware/dnstap/msg/msg.go
Comment (miekg) [inline middleware/dnstap/msg/msg.go]: Drop `Client`, it's cleaner.

File: middleware/dnstap/msg/msg.go
Comment (miekg) [inline middleware/dnstap/msg/msg.go]: Drop `Client`, it's cleaner.

File: middleware/dnstap/msg/util.go
Comment (miekg) [inline middleware/dnstap/msg/util.go]: Isn't it somewhat quaranteed the a `.RemoteAddr()` host the correct port ?

File: middleware/dnstap/msg/util.go
Comment (miekg) [inline middleware/dnstap/msg/util.go]: some here, you *have* a tcp connection... that can only come from a legal IP address, no?

File: middleware/dnstap/setup.go
Comment (miekg) [inline middleware/dnstap/setup.go]: We should probably talk about how the *dnstap* directive should look in the Corefile. Can you send a proposal?

**Review by miekg — CHANGES_REQUESTED** [review]
I'm proposing some method signature changes, which would impact any tests you'll write.

File: middleware/dnstap/msg/msg.go
Comment (2buisson) [reply middleware/dnstap/msg/msg.go]: I did this because there could also be `ToResolverResponse` or `ToAuthResponse` in fact one per [event type](https://godoc.org/github.com/dnstap/golang-dnstap#Message_Type), we don't plan on adding these I could remove `Client` yes, your call.

This is the alternative to `ToMsg`: a funcion per type, I wrote both to see what's best.

File: middleware/dnstap/msg/msg.go
Comment (2buisson) [reply middleware/dnstap/msg/msg.go]: I was thinking to combine `networkFromWriter` with for exemple `socketInfoFromResponse`, probably too much boilerplate.

File: middleware/dnstap/msg/util.go
Comment (2buisson) [reply middleware/dnstap/msg/util.go]: I guess, what does this mean for this function?

File: middleware/dnstap/msg/util.go
Comment (2buisson) [reply middleware/dnstap/msg/util.go]: Oh I see, if so I should take a `ResponseWriter` instead of the `net.Addr` and let it panic.

File: middleware/dnstap/msg/msg.go
Comment (2buisson) [reply middleware/dnstap/msg/msg.go]: Will do.

File: middleware/dnstap/writer.go
Comment (2buisson) [reply middleware/dnstap/writer.go]: Yes this is the option we discussed, it's also the last shady `true` parameter in the directive parameters.

File: middleware/dnstap/writer.go
Comment (2buisson) [reply middleware/dnstap/writer.go]: I tend to over do it, but in some case it adds a usefull context to errors.

File: middleware/dnstap/writer.go
Comment (2buisson) [reply middleware/dnstap/writer.go]: Found a way to do that.

File: middleware/dnstap/handler.go
Comment (2buisson) [reply middleware/dnstap/handler.go]: The epoch value of the query request, without this method, the info would be lost when `ResponseWriter.Write` is called and logs both the query and response.

File: middleware/dnstap/handler.go
Comment (2buisson) [reply middleware/dnstap/handler.go]: Will rename to Dnstap.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: middlewa..." ]
>+	return errors.Wrap(
>
>I tend to over do it, but in some case it adds a usefull context to errors.

Just return early with then error, you won't need errors.Wrap then.

/Miek

--
Miek Gieben

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: middlewa..." ]
>Yes this is the option we discussed, it's also the last shady `true` parameter in the directive options.

Ok. We may need to find something more descriptive than 'true' as a keyword in 
the config.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: middlewa..." ]
>varyoo commented on this pull request.
>
>I did this because there could also be `ToResolverResponse` or `ToAuthResponse` in fact one per [event type](https://godoc.org/github.com/dnstap/golang-dnstap#Message_Type), we don't plan on adding these I could remove `Client` yes, your call.
>
>This is the alternative to `ToMsg`: a funcion per type, I wrote both to see what's best.

Ah, ok, sgtm - leave it like it is.

File: middleware/dnstap/msg/util.go
Comment (2buisson) [reply middleware/dnstap/msg/util.go]: Or maybe a type assertion:

~~~
if addr, ok := conn.RemoteAddr().(*net.TCPAddr); ok {
    fmt.Println(addr.IP.String())
}
~~~

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: middlewa..." ]
>varyoo commented on this pull request.
>
>> +	}
>+	return uint32(i), nil
>+}
>+
>+func ipPort(a net.Addr) (ip []byte, port uint32, err error) {
>+	rawip, rawport, err := net.SplitHostPort(a.String())
>+	if err != nil {
>+		err = errors.Wrap(err, "split ip:port")
>+		return
>+	}
>+	port, err = parsePort(rawport)
>+	if err != nil {
>+		err = errors.Wrap(err, "port")
>+		return
>+	}
>+	netip := net.ParseIP(rawip)
>
>Or maybe a type assertion:
>
>~~~
>if addr, ok := conn.RemoteAddr().(*net.TCPAddr); ok {
>    fmt.Println(addr.IP.String())
>}
>~~~

Let's recap. You need the string (well bytes actually) of the remote IP so you 
can put that in the proto for dnstap? The above would work, but, I think you'll
need another type asserting to get the IP address (this would then work for TCP 
and UDP).

if addr, ok := conn.RemoteAddr().(*net.TCPAddr); ok {
    if ipaddr, ok := addr.(net.IPAddr); ok {
        fmt.Println(ipaddr.IP.String())
    }
}

(not: not tested)


/Miek

--
Miek Gieben

**Comment by johnbelamaric** [comment]
@varyoo what's the latest status here? I haven't had time to review but I may ask some other folks to take a look.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: middlewa..." ]
>@varyoo what's the latest status here? I haven't had time to review but I may ask some other folks to take a look.

I believe status is: "ready for another review"

would be good to let someone else than me take a look.

**Comment by 2buisson** [comment]
>I believe status is: "ready for another review"

Yes, the middleware is responsible to log client query and responses, it has been done and tested. There is still other [events](https://godoc.org/github.com/dnstap/golang-dnstap#Message_Type) to log to dnstap, but I think it is to be done outside of this middleware *(happy to implement these)*:

~~~
proxy . 192.168.1.1:53 {
    dnstap /tmp/dnstap.sock
}
~~~

Unless a review :blush: tell me otherwise, I don't have more plans regarding this issue.

**Comment by pmoroney** [comment]
I'm currently at GopherCon, but I can do a review on Saturday during the
community day if that's alright.

**Comment by johnbelamaric** [comment]
@varyoo awesome! why are you thinking those other events should be out of this middleware? As opposed to optionally enabled in this middleware?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: middlewa..." ]
>I'm currently at GopherCon, but I can do a review on Saturday during the
>community day if that's alright.

sgtm! Thank you

**Comment by 2buisson** [comment]
>why are you thinking those other events should be out of this middleware? As opposed to optionally enabled in this middleware?

If the proxy middleware is enabled, it will eventually call remote servers, those requests can be logged by dnstap, and they are invisble to the dnstap middleware. So why enable them under the dnstap middleware, since the the actual code can't be in it?

**Comment by johnbelamaric** [comment]
I see. We have a similar thing with tracing, where we want to utilize the Tracer object to push data from other middleware. The dnstap middleware will have all the config info; we would want to re-use that any any sockets it keeps open to push the data from the other middleware.

**Comment by johnbelamaric** [comment]
Is there a reason you are using `pkg/errors`? We don't use it anywhere else, we just use `errors` and `fmt`. Prefer not to add a dependency.

**Comment by johnbelamaric** [comment]
You need to add the dnstap entry to `middleware.cfg`. I thought I saw it in an earlier version of the PR but it's not in this one.

**Comment by johnbelamaric** [comment]
If I don't run the `dnstap` utility first, then I get this message. Instead, it should WARN and then have a goroutine that periodically checks if the socket exists. Or it could create the socket.

What happens if the pipe fills up? We should not block, but perhaps we should warn in this case.

```
ANN-M-JBELAMARIC:coredns jbelamaric$ ./coredns -conf Corefile-erratic
2017/07/15 10:13:51 middleware/dnstap: output: dial unix /etc/dnstap.sock: connect: no such file or directory
ANN-M-JBELAMARIC:coredns jbelamaric$
```

**Comment by 2buisson** [comment]
>The dnstap middleware will have all the config info; we would want to re-use that any any sockets it keeps open to push the data from the other middleware.

~~~Go
t := dnsserver.GetMiddleware(c, "trace")
P := &Proxy{Trace: t}
~~~

~~~Go
if p.Trace != nil {
        if t, ok := p.Trace.(trace.Trace); ok {
~~~

There is something else that can be done with a socket: the dnstap command line tool dosen't **yet** *(maybe the C version does)*, but could accept multiple connections! 

The following would be *then* completely transparent to the end user:

~~~Go
dnstap /tmp/dnstap.sock
proxy . 192.168.1.1:53 {
    dnstap /tmp/dnstap.sock
}
~~~

**Comment by johnbelamaric** [comment]
Travis build is failing because dnstap packages aren't in the vendor dir. If you add those too (it's in one of the READMEs) then we can see if Travis passes.

File: middleware/dnstap/msg/msg.go
Comment (pmoroney) [inline middleware/dnstap/msg/msg.go:17]: Is there a reason we are not just using the tap.Message struct here? It seems that we are using a subset of that struct.

Also, if we keep the Data struct, we may want to have all the functions below actually be methods of `*Data` since that is the first argument for each of them. i.e. `func (d *Data) FromRequest(r request.Request) error {`

File: middleware/dnstap/README.md
Comment (pmoroney) [inline middleware/dnstap/README.md:20]: Should we have something that periodically flushes the buffer, so messages aren't queued for long periods of time on low volume servers? I sent one query and it only flushed to dnstap when I killed coredns an hour later.

**Review by pmoroney — CHANGES_REQUESTED** [review]
Sorry for the delay. 
Besides the two questions I mentioned inline, it works for me.
Although, you should add comments to all exported types and functions. You can get a list by running `golint github.com/coredns/coredns/middleware/dnstap/...`. If you don't have golint, it can be gotten here: `github.com/golang/lint`

Thanks!

File: middleware/dnstap/msg/msg.go
Comment (2buisson) [reply middleware/dnstap/msg/msg.go:17]: Yes, in dnstap command-line tool's code, the combinaison of fields displayed for each event type falls under the fields described in `msg.Data`. And it is handy because [tap.Message](https://godoc.org/github.com/dnstap/golang-dnstap#Message) fieds are only pointers.

Argeed for the methods.

File: middleware/dnstap/README.md
Comment (2buisson) [reply middleware/dnstap/README.md:20]: Yes it could flushed periodically, the framestream library hides the buffer size.

**Comment by johnbelamaric** [comment]
@varyoo, let's move the proxy additions to a separate PR. I'd like to get basic `dnstap` support in, and that could delay it.

I think the only things to address are:

1) Don't fail startup if the socket doesn't exist. Instead either create it or WARN and periodically check if it exists.
2) Periodically flush the buffer.

@coredns/maintainers please chime in if you disagree.

**Comment by pmoroney** [comment]
@johnbelamaric I agree about splitting that to a new PR.
And on top of your things to address, comments should be placed on all exported types and functions so golint stops complaining.

**Review by miekg — APPROVED** [review]

**Comment by miekg** [comment]
I haven't looked in detail, but other have, so lgtm from my original review.

**Comment by miekg** [comment]
@varyoo this is not WIP anymore I think.
@pmoroney OK with merging?

**Comment by pmoroney** [comment]
Yep, looks good to me

**Comment by johnbelamaric** [comment]
👍 nice, @varyoo 

you should now rebase the other PR :)

**Comment by alkino** [comment]
nice @varyoo 
But miekg you change the title of PR and not the title of commit. It is written "WIP" in history now. Too late but sad ^^

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] middleware/dn..." ]
>nice @varyoo
>But miekg you change the title of PR and not the title of commit. It is written "WIP" in history now. Too late but sad ^^

ah well :)

**Pull request merged**
